### PR TITLE
Fix "No notifications after update 2.9 and 2.9.1" (#5535)

### LINKED ISSUE #5535: No notifications after update 2.9 and 2.9.1
Hey :)
After tue 2.9 Update, my S8 does not display a notification reminder for the remaining cards anymore. I have since deinstalled and reinstalled and checked the settings and permissions and all seems in order. However, the notifications have not reappeared :)
Thank you for any help!

###### Reproduction Steps

1. 
2. 
3. 

###### Debug info
Refer to the [support page](https://ankidroid.org/docs/help.html) if you are unsure where to get the "debug info".

###### Research
*Enter an [ x ] character to confirm the points below:*

[ x ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid

[ x ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue

[ x ] I have searched for similar existing issues here and on the user forum

### PR DESCRIPTION
**Pull request by tda1009**
## Purpose / Description
An attempt to fix issue "No notifications after update 2.9 and 2.9.1" #5535 

## Fixes
Fix #5535 

## Approach
From my experience with AnkiDroid, there are 2 types of notifications  in the app:
(1) The general notification which has the form "%d AnkiDroid cards due"  
(2) The reminder (specific to each deck) which has the form "%d cards to review in <DeckName>"  
For convenience, in the following text, I regard (1) as notification, and (2) as reminder.  
 
I ran some tests. It turned out that the notification only works when the widget is enabled.

Steps to reproduce this phenomenon:
1. Install AnkiDroid.
2. Download and import a shared deck with more than 10 cards.  
3. Enable widget (the AnkiDroid Small Widget).
(by "enable widget" I mean add the AnkiDroid's widget onto the Home Screen of the phone)
4. Choose Settings -> AnkiDroid General Settings -> Notifications -> Notify When More than 10 cards due.

If all of these steps are done, then AnkiDroid shows the notification (*).
But if we omit step 3 (which means the widget is not enabled), then AnkiDroid does not show the notification.

(*) We might have to wait some minutes before the notification is shown.
  
The above mentioned behavior is unexpected, since AnkiDroid should show the notification regardless of the widget's state.  
In my opinion, the normal behaviour should be: After the user chooses a certain due threshold in Settings, and if the user has more due cards than this threshold, the notification should be displayed. The user doesn't have to enable the widget. 
 
I see 2 problems in the code which are probably the culprits.

Problem 1:
Class NotificationService uses a method of class WidgetStatus to get the due value.  
The problem is, this method only works correctly if the widget is enabled.  
if widget is enabled, this method returns the correct due value.  
if widget is not enabled, this method always returns 0.  
One solution to this (I think) is to allow NotificationService to get the due value directly from Collection.  
 And since getting due from Collection might not be quick, NotificationService should use an AsyncTask to do the job. 

Problem 2:
In class BootService, method scheduleNotification(), statement if, the comparison operator is currently <=.   
I think it should be >=  .

The patch I provided in this PR is an attempt to solve our notification problem. It implements my aforementioned suggestions.

## How Has This Been Tested?

I tested this patch in both the emulator (API 28) and a physical phone, and AnkiDroid worked as expected (ie. the notification, once enabled in Settings, was shown correctly regardless of the state of the widget).  
I also ran all unit tests and instrumentation tests (on emulator) and this patch didn't break any of them.  

## Checklist
_Please, go through these checks before submitting the PR._

- [x ] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x ] You have a descriptive commit message with a short title.
- [x ] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x ] You have commented your code, particularly in hard-to-understand areas
- [x ] You have performed a self-review of your own code

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/services/NotificationService.java [hunk 1]
```diff
@@ -21,63 +21,116 @@
...
-                    PendingIntent.FLAG_UPDATE_CURRENT);
-            builder.setContentIntent(resultPendingIntent);
-            // mId allows you to update the notification later on.
-            manager.notify(WIDGET_NOTIFY_ID, builder.build());
-        } else {
-            // Cancel the existing notification, if any.
-            manager.cancel(WIDGET_NOTIFY_ID);
+            dueCardsCount = total[0] + total[1] + total[2];
+            return dueCardsCount;
         }
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/services/NotificationService.java]: I don't see the need for the 'total' temporary array as we never handle the counts in a decomposed way, seems like we could just increment 'dueCardsCount' directly then return it?

```java
            int dueCardsCount = 0;
            Collection col = CollectionHelper.getInstance().getCol(context);
            // Ensure queues are reset if we cross over to the next day.
            col.getSched()._checkDay();
            // Only count the top-level decks in the total
            List<Sched.DeckDueTreeNode> nodes = col.getSched().deckDueTree();
            for (Sched.DeckDueTreeNode node : nodes) {
                dueCardsCount += node.newCount + node.lrnCount + node.revCount;
            }
            return dueCardsCount;
```

**Comment by mikehardy** [comment]
@tda1009 both of these changes look right, thank you for looking into this and fixing it! Was definitely an irritation for users on the 2.8->2.9 transition

I left one comment I think can simplify the due card logic, other than that I don't see anything wrong with this currently. I'm going to re-read it a couple times and compare with the widget, but other than the due card calc I'm +1 on it

**Comment by mikehardy** [comment]
On re-read I was chasing a second nagging thought - one thing in the feedback on the missing notification was a mention that the notification updated during reviews so it was always up to date. That's a use case that maybe wasn't tested here? Examining the notification (without dismissing it) as you do reviews?

This mechanism is what is doing it for WidgetStatus, the static update method here: https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/widget/WidgetStatus.java#L54

Called any time something affecting counts happens:

```
(base) mike@kunashir:~/work/AnkiDroid/Anki-Android/AnkiDroid/src/main/java (dependency-updates) % grep -r WidgetStatus.update *
com/ichi2/widget/AnkiDroidWidgetSmall.java:        WidgetStatus.update(context);
com/ichi2/widget/AnkiDroidWidgetSmall.java:                                    WidgetStatus.update(getBaseContext());
com/ichi2/widget/WidgetStatus.java:            Timber.d("WidgetStatus.update(): updating");
com/ichi2/widget/WidgetStatus.java:            Timber.d("WidgetStatus.update(): already running or not enabled");
com/ichi2/anki/NoteEditor.java:            WidgetStatus.update(this);
com/ichi2/anki/ModelFieldEditor.java:            WidgetStatus.update(this);
com/ichi2/anki/Reviewer.java:            WidgetStatus.update(this);
com/ichi2/anki/DeckPicker.java:            WidgetStatus.update(this);
com/ichi2/anki/DeckPicker.java:                WidgetStatus.update(DeckPicker.this);
com/ichi2/anki/StudyOptionsActivity.java:            WidgetStatus.update(this);
com/ichi2/anki/ModelBrowser.java:            WidgetStatus.update(this);
com/ichi2/anki/CardBrowser.java:            WidgetStatus.update(this);

```

I don't see anything in the NotificationService that does that currently so if it is not using the widget status for counts *or* the widget is not enabled, then the counts will go stale

As a design to have the notification count updated I think

- the NotificationService will need some static hook to update things e.g. NotificationService.update(Context)
- having two static functions doing the same both called from throughout the code is silly, so there should be only one call
- the notifications are used more than the widget so it should be the call in all the other code I think, and then it should chain out and call WidgetStatus.update 

I don't think there is more to it than that?

**Comment by mikehardy** [comment]
A further thought as I ponder this: the WidgetStatus AsyncTask and the one proposed here are essentially equivalent, it's just the result that is handle differently - the onPostExecute. The logic should probably be extracted to single object with an abstract onPostExecute, but a concrete implementation of the deck due counter.

Or maybe the whole thing could just be renamed to "CardsDueStatus", and it could both check for widget / update Widget && check for notifications enabled + over threshold / update notification like this chunk https://github.com/mikehardy/rnfbdemo/blob/master/make-demo-v6.sh#L54

Basically just trying to reduce duplication of AsyncTask logic checking due counts - it feels like an error to have that implemented almost identically in two places

**Comment by tda1009** [comment]
Hi @mikehardy , thank you for your responses.   
   
Today I created another commit, and in this commit I addressed the problems that you've pointed out .
   
(This commit is a big change from my previous commit. However, it is a smaller change from the latest commit of upstream repo. So it is easier to understand things if we compare this commit to the upstream repo's commit.)
   
Regarding the problem "notification is not updated during reviews" I ran some tests in Emulator and the tests confirmed that in AnkiDroid version 2.8, notification is updated, but in version 2.9, it is not. In other words, this problem only exists in AnkiDroid 2.9, not in 2.8.  
   
I found out that in AnkiDroid 2.9 this problem is caused by a change in PR #4697 "Show notifications even when app is in background" (13.09.2017)
Specifically, in this PR, file WidgetStatus.java, method onPostExecute(), the code which starts NotificationService was removed.   
When I placed the code back into its original position, this problem disappeared (ie. notification was updated during reviews again).  
   
As to why the code was removed, according to the author of this PR:  
"Regarding the widget, it was starting the NotificationService on every update. Now that the service is started automatically via the BootService, there's no reason to start it from the widget any more."  
Unfortunately, in reality as we've all seen, this removal caused notification not updated during reviews.  
Based on the author's statement, I think that adding the code back does no harm to his PR, therefore in this PR I did just that.  
Also, while making this change I noticed that the type of NotificationService was changed from Service to BroadcastReceiver (in PR #4928 on 05.09.2018), so in this commit I registered NotificationService in method onCreate() of class AnkiDroidApp.  
   
Besides, in order to avoid the code duplication problem you mentioned earlier, and to keep code changes to minimum, in this commit I used again the fetchDue() method from WidgetStatus to get due value for NotificationService. This means no more AsyncTask in class NotificationService. In other words, now I virtually don't touch class NotificationService. It is essentially the same as it was in the last commit of the upstream repo.  
   
I also modified the boolean expression of the if statement in method WidgetStatus.update(), so that the notification is shown even when the widget is not enabled. This modification is based on what AnkiDroid 2.8 did.  
   
After the above changes, the notification is now shown even when the widget is not enabled, and the notification is also updated in real-time (while reviewing).  
   
After making these changes, I ran the unit tests and instrumentation tests again, and 8/52 unit tests were failed.  
I then updated Android Gradle Plugin from 3.5.2 to 3.5.3, and all of the tests were passed.  
 
 
   
Please take a look at this commit to see if it is ok. Thanks.

**Comment by mikehardy** [comment]
@tda1009 I've reviewed this and it seems to make sense, I'd like to merge it in
CI was broken because of #5649 but I just fixed it with #5657 - could you please rebase against master and re-push (it'll be a force-push I think) your branch so the PR builds with the updated CI config? It should go green then 

Sorry for the hassle - the emulator had a breaking change and it took a little bit of time to sort out

**Comment by tda1009** [comment]
Hi @mikehardy , as per your request I already rebased this PR on top of upstream/master and re-push.

I also restructured the PR. In the new structure, each commit performs an atomic task:

Commit 1: Update Android Gradle Plugin from 3.5.2 to 3.5.3
Commit 2: Make notification appear even when widget is not enabled
Commit 3: Make notification update during reviews

I think this will make things easier to follow.

File: AnkiDroid/src/main/java/com/ichi2/anki/services/BootService.java [hunk 2]
```diff
@@ -89,7 +89,7 @@ private void scheduleDeckReminder(Context context) {
     public static void scheduleNotification(Context context) {
         AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
         SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(context);
-        if (Integer.parseInt(sp.getString("minimumCardsDueForNotification", "1000001")) <= 1000000) {
+        if (Integer.parseInt(sp.getString("minimumCardsDueForNotification", "1000001")) >= 1000000) {
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/services/BootService.java]: Have you tested the code without this change? On first glance, my reaction is that the change doesn't look correct to me, because we use some awkward special values for disabling notifications in our preferences.

Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/services/BootService.java]: The submitter did test, plus if you re-inspect you can see it was exactly wrong before, My previous edit had the test incorrect. The goal is to express the idea "if the configured minimum cards due is more than a ridiculous number (our special value that flags notifications as disabled - 1,000,000), then don't schedule anything".

What the code used to say was "return / don't schedule anything if it's less than the special value", which is wrong on the face of it, and with this change it continues on and schedules the notification as it should.
